File: artifacts/tulipindicators/decompose/src/dm.c
int ti_dm_start(const double *options)
{
  return ((int) options[0]) - 1;
}


/*** DEPENDENCIES:
***/


// hint:  ['plus_dm_idx_ref is a mutable refrence to unsigned int', 'minus_dm_idx_ref is a mutable refrence to unsigned int', 'dmup_ref is a mutable refrence to double', 'dmdown_ref is a mutable refrence to double']
void helper_ti_dm_1(unsigned int * const plus_dm_idx_ref, unsigned int * const minus_dm_idx_ref, double * const dmup_ref, double * const dmdown_ref, const double * const high, const double * const low, double * const plus_dm, double * const minus_dm, const double per, int i)
{
  unsigned int plus_dm_idx = *plus_dm_idx_ref;
  unsigned int minus_dm_idx = *minus_dm_idx_ref;
  double dmup = *dmup_ref;
  double dmdown = *dmdown_ref;
  double dp;
  double dm;
  do
  {
    dp = high[i] - high[i - 1];
    dm = low[i - 1] - low[i];
    if (dp < 0)
    {
      dp = 0;
    }
    else
      if (dp > dm)
    {
      dm = 0;
    }
    if (dm < 0)
    {
      dm = 0;
    }
    else
      if (dm > dp)
    {
      dp = 0;
    }
  }
  while (0);
  dmup = (dmup * per) + dp;
  dmdown = (dmdown * per) + dm;
  plus_dm[plus_dm_idx] = dmup;
  plus_dm_idx += 1;
  *((&minus_dm[minus_dm_idx])++) = dmdown;
  *plus_dm_idx_ref = plus_dm_idx;
  *minus_dm_idx_ref = minus_dm_idx;
  *dmup_ref = dmup;
  *dmdown_ref = dmdown;
}


/*** DEPENDENCIES:
***/


int ti_dm(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *high = inputs[0];
  const double *low = inputs[1];
  const int period = (int) options[0];
  double *plus_dm = outputs[0];
  unsigned int plus_dm_idx = 0;
  double *minus_dm = outputs[1];
  unsigned int minus_dm_idx = 0;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_dm_start(options))
  {
    return 0;
  }
  const double per = (((double) period) - 1) / ((double) period);
  double dmup = 0;
  double dmdown = 0;
  int i;
  for (i = 1; i < period; i += 1)
  {
    double dp;
    double dm;
    do
    {
      dp = high[i] - high[i - 1];
      dm = low[i - 1] - low[i];
      if (dp < 0)
      {
        dp = 0;
      }
      else
        if (dp > dm)
      {
        dm = 0;
      }
      if (dm < 0)
      {
        dm = 0;
      }
      else
        if (dm > dp)
      {
        dp = 0;
      }
    }
    while (0);
    dmup += dp;
    dmdown += dm;
  }

  plus_dm[plus_dm_idx] = dmup;
  plus_dm_idx += 1;
  minus_dm[minus_dm_idx] = dmdown;
  minus_dm_idx += 1;
  for (i = period; i < size; i += 1)
  {
    helper_ti_dm_1(&plus_dm_idx, &minus_dm_idx, &dmup, &dmdown, high, low, plus_dm, minus_dm, per, i);
  }

  assert(((&plus_dm[plus_dm_idx]) - outputs[0]) == (size - ti_dm_start(options)));
  assert(((&minus_dm[minus_dm_idx]) - outputs[1]) == (size - ti_dm_start(options)));
  return 0;
}


/*** DEPENDENCIES:
void helper_ti_dm_1(unsigned int * const plus_dm_idx_ref, unsigned int * const minus_dm_idx_ref, double * const dmup_ref, double * const dmdown_ref, const double * const high, const double * const low, double * const plus_dm, double * const minus_dm, const double per, int i)
{
  unsigned int plus_dm_idx = *plus_dm_idx_ref;
  unsigned int minus_dm_idx = *minus_dm_idx_ref;
  double dmup = *dmup_ref;
  double dmdown = *dmdown_ref;
  double dp;
  double dm;
  do
  {
    dp = high[i] - high[i - 1];
    dm = low[i - 1] - low[i];
    if (dp < 0)
    {
      dp = 0;
    }
    else
      if (dp > dm)
    {
      dm = 0;
    }
    if (dm < 0)
    {
      dm = 0;
    }
    else
      if (dm > dp)
    {
      dp = 0;
    }
  }
  while (0);
  dmup = (dmup * per) + dp;
  dmdown = (dmdown * per) + dm;
  plus_dm[plus_dm_idx] = dmup;
  plus_dm_idx += 1;
  *((&minus_dm[minus_dm_idx])++) = dmdown;
  *plus_dm_idx_ref = plus_dm_idx;
  *minus_dm_idx_ref = minus_dm_idx;
  *dmup_ref = dmup;
  *dmdown_ref = dmdown;
}


----------------------------
int ti_dm_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
***/

File: artifacts/tulipindicators/translate_new_ablation/src/dm.rs
use crate::*;
use std::assert;

pub fn ti_dm_start(options: Option<&[f64]>) -> Option<i32> {
    let options = options?;
    if options.is_empty() {
        return None;
    }
    Some(options[0] as i32 - 1)
}
pub fn ti_dm(size: i32, inputs: &[&[f64]], options: &[f64], outputs: &mut [&mut [f64]]) -> i32 {
    let high = inputs[0];
    let low = inputs[1];
    let period = options[0] as i32;
    
    if period < 1 {
        return 1;
    }
    
    let start_index = match ti_dm_start(Some(options)) {
        Some(idx) => idx,
        None => return 0,
    };
    
    if size <= start_index {
        return 0;
    }
    
    let per = ((period as f64) - 1.0) / (period as f64);
    let mut dmup = 0.0;
    let mut dmdown = 0.0;
    
    // Use split_at_mut to get non-overlapping mutable slices
    let (plus_dm_slice, rest) = outputs.split_at_mut(1);
    let plus_dm = &mut plus_dm_slice[0];
    let minus_dm = &mut rest[0];
    
    let mut plus_dm_idx = 0;
    let mut minus_dm_idx = 0;

    for i in 1..period {
        let (dp, dm) = {
            let mut dp = high[i as usize] - high[(i - 1) as usize];
            let mut dm = low[(i - 1) as usize] - low[i as usize];
            
            if dp < 0.0 {
                dp = 0.0;
            } else if dp > dm {
                dm = 0.0;
            }
            
            if dm < 0.0 {
                dm = 0.0;
            } else if dm > dp {
                dp = 0.0;
            }
            
            (dp, dm)
        };
        
        dmup += dp;
        dmdown += dm;
    }
    
    plus_dm[plus_dm_idx] = dmup;
    plus_dm_idx += 1;
    minus_dm[minus_dm_idx] = dmdown;
    minus_dm_idx += 1;
    
    for i in period..size {
        let (dp, dm) = {
            let i_usize = i as usize;
            let mut dp = high[i_usize] - high[(i - 1) as usize];
            let mut dm = low[(i - 1) as usize] - low[i_usize];
            
            if dp < 0.0 {
                dp = 0.0;
            } else if dp > dm {
                dm = 0.0;
            }
            
            if dm < 0.0 {
                dm = 0.0;
            } else if dm > dp {
                dp = 0.0;
            }
            
            (dp, dm)
        };
        
        dmup = (dmup * per) + dp;
        dmdown = (dmdown * per) + dm;
        
        plus_dm[plus_dm_idx] = dmup;
        plus_dm_idx += 1;
        minus_dm[minus_dm_idx] = dmdown;
        minus_dm_idx += 1;
    }
    
    let plus_dm_written = plus_dm_idx;
    let minus_dm_written = minus_dm_idx;
    let expected = (size - start_index) as usize;
    
    assert!(plus_dm_written == expected);
    assert!(minus_dm_written == expected);
    
    0
}
